### ISSUE TITLE
[setters] core actions

### ISSUE THREAD

**Issue by louisameline** [vuex-cloud/rfcs#4]
What API to handle 1-way sync and other stuff that needs parameters?

Sometimes (or maybe always for some users) I will want to make sure the data is saved to Firestore before adding it to my local store.

These are the default calls:
```javascript
// this creates
dispatch('module/insert', { a: { bb: 1, cc: 1 }})

// this overwrites the entire doc, there is no "partial override"
dispatch('module/replace', { a: { dd: 1}})

// this replaces `a` with an empty object
dispatch('module/assign', { a: {}})

// you may use the dot notation to replace a nested object instead of extending it
dispatch('module/assign', { 'a.bb.ccc': { ...newSetOfProperties } })

// this updates `a.bb` and adds `a.dd`. `a.cc` is left intact
dispatch('module/merge', { a: { bb: 2, dd: 2 }})

// this deletes a property
dispatch('module/delete', 'a')

// this deletes the document
dispatch('module/delete')
```

Now how to pass other parameters, for example to specify if we want 1-way or 2-way sync?
No matter the action, I think we should say:
> If there is a root property named `parameters` at the root of the object, then we will search for a `data` property instead of using the root properties. 

```javascript
// `optimistic: false` for 1-way, `optimistic: true` for 2-way
dispatch('module/assign', {
  data: { a: { bb: 1, cc: 1 }},
  parameters: { optimistic: true }
)
```

I'm not sure we should allow the creation of a document by calling an action on a collection anymore (if we can do it another way), but for the sake of the argument, if we did it would be like:
```javascript
dispatch('collectionModule/createChild', {
  id: 'docId',
  data: { a: { bb: 1, cc: 1 }},
  parameters: { optimistic: true }
})
```
If feasible at some point, I'd rather have directly
```javascript
dispatch('collectionModule/docId/create', {
  data: { a: { bb: 1, cc: 1 }},
  parameters: { optimistic: true }
})
```
Note: maybe the name "parameters" will be an issue for some people if they already have a property by that name, and it should be configurable.

**Comment by louisameline** [comment]
Luca, I let things sink in after our conversation.

1. I now agree with you that the object notation should be presented as the default way of doing things, for clarity.
2. However I do think we need the dot notation and I think that I will personnaly use it more often than the object notation.
3. I think that the `patch` action should actually be called `extend`, it makes so much sense, it makes really obvious what it does. And so `update` can actually be used to replace parts of a document, like Firebase does. It all falls into place!

Check out the original post above, I have updated it with my latest thoughts on how to lay things out to the users. I hope you'll like it

To recap the arguments in favor of the dot notation that I do hope to see implemented, I would say this:
1. **Allows partial object replacement.**
Even if 5% of users asked for it, it's still a lot of people and I see no reason not to support it
2. **Easier to write and read**
```javascript
dispatch('module/update', { 'my.nested.property': 123 })
```
rather than
```javascript
dispatch('module/extend', { my: { nested: { property: 123 }}})
```
When I see the dot notation that uses a single string, I immediately know there is only 1 property updated and it's clear which one. My brain doesn't have to actually conceptualize an object. Do I sound super lazy? xp

**Comment by louisameline** [comment]
Fun fact: I just found a bug in my app a few seconds ago, I had overlooked the fact that `patch` would not delete an old prop in an objected I updated ^^'

I can't conceive two calls (one to set null/delete and one to repatch) because it's both awkward and more expensive on my bills. Who knows what kind of errors can happen anyway when you batch two write operations on the same document and same property, that's playing with fire...

**Comment by mesqueeb** [comment]
I'll meet you in the middle. 😄 

I understand that `extend` is usually used for extending a class, but I don't like the way it sounds, I don't have a good explanation though... 😅 

What do you think about these two options:
- keep calling it `patch`: Psychologically it'll be easier for all current v-e-f devs, since patch stays as it is
- call it `merge`: can't get more literal than this right? 😆 

I can agree with implementing `update` as a new action similar to the Firestore SDK.

```js
// this creates
dispatch('module/insert', { a: { bb: 1, cc: 1 }})

// this overwrites the entire doc, there is no "partial override"
dispatch('module/replace', { a: { dd: 1}})

// this updates `a.bb` and adds `a.dd`. `a.cc` is left intact
dispatch('module/patch', { a: { bb: 2, dd: 2 }})
// OR
dispatch('module/merge', { a: { bb: 2, dd: 2 }})

// this replaces `a` with an empty object
dispatch('module/update', { a: {}})

// you may use the dot notation to replace a nested object instead of extending it
dispatch('module/update', { 'a.bb.ccc': { ...newSetOfProperties } })
```

Now we need to think about wether or not we will use a module per doc or not. I have some ideas that I will write out today.

**Comment by louisameline** [comment]
Hi Luca, in JavaScript you "merge" arrays, but you "extend" objects. It's actually the proper technical term you know ;) I thought I'd tell you when I saw that your merge-anything fonction had the wrong name haha ^^ Think of jQuery's extend method, maybe you're too young to have used it but it's what everybody used for 10 years to "merge" objects. I wouldn't say that I don't understand what "merge" means when I see it about objects, but I definitely think to myself "this person is not using the right vocabulary".

**Comment by louisameline** [comment]
`module/assign` would work instead of `module/extend` if it suits you better.
But since `assign` is so much associated to `Object.assign` which is shallow (non-recursive), by default the action should be shallow too: so it would replace `module/update`.
Additionally, it could have a `deep` parameter.
So, the options:

```javascript
// assigns (shallow)
dispatch('module/update', { a: { bb: 2, dd: 2 }})
// extends (deep)
dispatch('module/extend', { a: { bb: 2, dd: 2 }})
```
versus
```javascript
// assigns (shallow)
dispatch('module/assign', { a: { bb: 2, dd: 2 }})
// extends (deep)
dispatch('module/assign', {
  data: { a: { bb: 2, dd: 2 }},
  parameters: { deep: true }
})
```
It's more cumbersome to have to provide a parameter, but on the bright side, there's only one action to remember.

We could also implement `module/assign` as suggested above AND also provide a handy `module/extend` alias:
```javascript
dispatch('module/extend', { a: { bb: 2, dd: 2 }})
// same as
dispatch('module/assign', {
  data: { a: { bb: 2, dd: 2 }},
  parameters: { deep: true }
})
```

**Comment by mesqueeb** [comment]
@louisameline 
I looked further into the matter of merge vs extend. And these were my findings:

### in jQuery
`jQuery.extend( target [, object1 ] [, objectN ] )`
is similar to Object.assign

`jQuery.extend( true, target [, object1 ] [, objectN ] )`
is similar to a deep Object.assign

`jQuery.merge( first, second )`
is used for arrays only

### in Angular

`angular.extend(dst, src1, src2, ...)`
is similar to Object.assign

`angular.merge(dst, src1, src2, ...)`
is similar to a deep Object.assign

([link](http://davidcai.github.io/blog/posts/copy-vs-extend-vs-merge/))

### in lodash

`lodash.assign` or `lodash.extend`
is similar to Object.assign

`lodash.merge`
is similar to a deep Object.assign

((link)[https://stackoverflow.com/questions/19965844/lodash-difference-between-extend-assign-and-merge])

### in plain JavaScript

> The `extends` keyword is used in class declarations or class expressions to create a class which is a child of another class.

([link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/extends))

----

- In both jQuery, Angular and lodash, `extend` seems to be closest to the shallow Object.assign.
- The usage of `merge` is a deep `Object.assign` in Angular and lodash, while jQuery uses it for arrays only.

I couldn't find any other prominent libraries providing functions called `extend` or `merge`, but if you know any others let me know.

---

Based on this information, I think I still prefer (1) rename the current `patch` action of v-e-f to `merge` or (2) keep it called `patch`.

Let me know your thoughts. 😄

**Comment by louisameline** [comment]
Without further ado, I think we can settle for "assign" as shallow and "merge" as deep.

I read some more yesterday too about how it's used today, and I feel that the meaning of the word "extend" has shifted through the years. Today people use it less about editing objects and more strictly for inheritance because of the evolving standards. Ok to follow the current :)

**Comment by louisameline** [comment]
Reminder for me to update the docs: we need to explain that an object which can be deleted by the app at some point should be modified with the `assign` action, not the `extend` action, because in case it has been deleted at Firestore and that we're not aware, `extend` will recreate the object with only some of its properties, which will probably generate errors. So `extend` should be used only when we know the object will be there when Firestore runs the write operation (a transaction can be used for that matter).

**Comment by mesqueeb** [comment]
Hey @louisameline I think we can close issues we agreed on, so we can check the [roadmap](https://github.com/vuex-cloud/rfcs/projects/1) for the topics left. 😉

**Comment by louisameline** [comment]
There's a reminder for me up there though. Maybe we can add a `settled` tag on issues and close only once implemented?

**Comment by louisameline** [comment]
What did you think of the
```javascript
dispatch('module/assign', {
  data: { a: { bb: 1, cc: 1 }},
  parameters: { optimistic: true }
)
```
notation? Makes sense so far? We'll have to see if it can be applied in all situations.

**Comment by mesqueeb** [comment]
### About optimistic UI (updating local store _before_ or _after_ server sync resolves)

I'm not 100% sure what you mean by 1- or 2-way sync. But if it's optimistic UI you're talking about then this is what I think:

I think we should opt for the library to **always have optimistic UI behaviour**:
1. the central (vuex) store is updated
2. the changes are synced to the server
3. in case the changes saw an error they should be reverted in the store

I don't see any reason not to use optimistic UI. I think this is one of the opinionated things about vuex-easy-firestore I'd like to keep.

In cases the dev needs to do extra things on errors, he can easily do so by setting up hooks.

### Devs who want server sync first → local store update after
In cases he _really_ wants to delay the vuex store being updated until AFTER the server is synced without errors, I'd say this:
- **if the dev wants this behaviour by default**, I'd recommend him/her to use another library or his own implementation, or fork our library and adjust
- **if the dev wants this behaviour one just a few calls here and there,** I'd just recommend in the docs how simple this is implemented via the native Firestore SDK.

### About the implementation of the actions
1. Please see my [post about the vuex-cloud Architecture](https://github.com/vuex-cloud/rfcs/issues/11#issuecomment-569467055) first
2. Once you understand my current thoughts, you'll understand that I think actions are to be dispatched like so (A or B below):

#### A. calling the VuexCloudModule instance natively
```js
import { VuexCloudModule as TodoTasks } from 'store/TodoTasks'

TodoTasks.insert({name: 'Luca'})
// or
TodoTasks.id('some-id-123').merge({age: 29})
```
> Extra config on a per-action-basis is to be passed as second parameter.

#### B. registering a vuex action for it

The dev can then choose to create vuex actions for these:
```js
// TodoTasks.js vuex module
export const VuexCloudModule = VuexCloud({ /* config */ })

export default {
  actions: {
    insert (context, payload) {
      VuexCloudModule.insert(payload)
      
      // or with extra config as per your example
      VuexCloudModule.insert(payload.data, payload.parameters)
    },
  },
}
```

**Comment by louisameline** [comment]
Having the methods directly accessible on the module object is good. Allowing this and going for events is pretty much the path I took for Tooltipster too, and it was much better afterwards.
However I think the actions should stay the primary way to work with the library, at least in our communication. At first, we don't want to break the habits of our regular Vuex users. I want the learning curve / transition to be as smooth as possible. One of our promises must be "keep working with Vuex pretty much as you're used and we'll take care of the sync". And for that, we'll need something handier than the mapping you proposed in `B. registering a vuex action`. I'll think about a syntax later today.

About 1-way sync: yes, I refer to optimistic UI. I call it 1-way sync because the data goes into a cyclical, single-direction flow: UI > Firestore SDK > Firestore > Vuex > UI. While optimistic UI / latency compensation updates both local and remote in parallel, with more complexities in case of remote failures.

 I do think 1-way sync is mandatory:
- you cannot build a library that aims to connect at all local stores and all remotes stores, but ignores completely this major pattern. If we say we're gonna be universal, we need to support universal patterns.
- Any serious application needs it at some point for critical things. The point is that some things are too important to act as if they were going to succeed. Sometimes you just need to be sure, and not let the user think it went ok until you know it didn't. Especially since you can have watchers that will trigger a whole lot of things, which in turn could trigger more errors if the first thing didn't go right, leaving you with a gigantic mess to revert by yourself. Try to debug that! The truth is, optimistic UI is only good for small things that can be reverted without hassle.
- saying "call Firestore SDK yourself" is a big no-no. Our promise is that the user doesn't have to worry about Firestore, he only deals with his local store. All data processing (think of guards, transforms, etc.) is already set up through our event callbacks (hooks if you will), you don't want to work with Firestore outside of vue-sync.
- By allowing 1-way sync, we make Vuexfire completely useless, because Vuexfire does 1-way sync only and leaves you with all the Firebase SDK boilerplate. Don't you want to deprecate Vuexfire and win back its 2600 Github stars? :)

**Comment by mesqueeb** [comment]
Thanks for your comment.

### About 1/2 way sync

I agree now we should offer both patterns:
- update local store → update remote store [→ revert local store on error]
- update remote store → update local store on success

However, calling this 1- or 2-way sync is very confusing. I believe both ways can be considered to be called "2-way sync" it's just a matter of which order to do this in. 😄

- 1-way sync is more about: only reading data (not writing) or only writing data (not reading)
- 2-way sync is more about: reading & writing data at the same time

But this is just my understanding, so if it's OK with you, when we start writing docs, let's find a good way to call our sync patterns which causes no confusion.

### About Vuex actions

I can agree that the current method, while I think it's best to be as explicit as possible, can be made a little bit easier for the devs.

My original plan however, was for the devs to be explicit about which actions are added to the actions of the vuex module. This is mainly because I want to avoid magic actions being added without them being easy to trace where they come from.

So, keeping that in mind, I believe my next suggestion might be _the golden mid-road_. (it's a dutch saying 😛 )

```js
// TodoTasks.js vuex module config file

// the VuexCloudModule instance
export const vuexCloudModule = VuexCloud.registerModule({ /* config */ })

const { insert, stream, get } = vuexCloudModule.vuexActions

// the vuex module options
export default {
  actions: {
    insert,
    stream,
    get,
    // other actions of the dev
  },
}
```

This way, the dev can still mouse over the actions to see what they do, but we have provided the boilerplate for them.

Even for dev who like to split modules, they can instanciate `vuexCloudModule` in the `actions.js` file so they can immediately retrieve the pre-made actions, or they can easily import the `vuexCloudModule` instance for the module into the `actions.js`.

As a cool bonus from retrieving the pre-made actions from the `vuexCloudModule` instance is that _because_ it's instanciated like so:
```js
export const vuexCloudModule = VuexCloud.registerModule({
  localStoreConfig: {path: 'TodoTasks'}
})
```
the core lib knows to use the plugin registered for the "local store", which is in this case Vuex, and thus **the logic of providing the boilerplate of the Vuex actions can be written in the Vuex plugin!**

**Comment by louisameline** [comment]
This sounds great :D

**Comment by louisameline** [comment]
For the record, as discussed today, this is a proposal:

```javascript
// the `set` action instead of `assign`. Overwrites whatever value is at the given path.
.dispatch('module/set', { path: value })

// works with strings as well as objects
.dispatch('module/set', { 'name.first': 'Lucas'})
// this completely overwrites the object that existed before (not a merge)
.dispatch('module/set', { 'name': { first: 'Lucas', last: 'Ban' }})
// potentially nested
.dispatch('module/set', { 'personalInfo.identity.name': { first: 'Lucas', last: 'Ban' }})
```

**Comment by louisameline** [comment]
My rant of the day: it's annoying to allow both these syntaxes:
```javascript
.dispatch('module/set', { ...someProperties })
```
and this :
```javascript
.dispatch('module/set', { data: { ...someProperties  }, parameters: { ...someParameters } })
```
Of course it's good to not mix the user properties and the options in a single object, but:
- the syntax is cumbersome
- if the user has properties named `parameters`, it's error-prone

If we keep a vuex-like `dispatch` method, so that VueSync feels familiar to Vuex users, I'd like to adapt it and allow multiple parameters:
```javascript
.dispatch('module/set', { stuff: 123 }, { ...myoptions })
```
I'll check what the reasons were to allow only a single parameter, but right now I wish we got rid of that.
